src/components/ha-target-picker.ts

```diff
@@ -0,0 +1,592 @@
+   * @type {Array}
+   * @attr include-domains
+   */
+  @property({ type: Array, attribute: "include-domains" })
+  public includeDomains?: string[];
+
+  /**
+   * Show only targets with entities of these device classes.
+   * @type {Array}
+   * @attr include-device-classes
+   */
+  @property({ type: Array, attribute: "include-device-classes" })
+  public includeDeviceClasses?: string[];
+
+  @property() public deviceFilter?: HaDevicePickerDeviceFilterFunc;
+
+  @property() public entityRegFilter?: (entity: EntityRegistryEntry) => boolean;
+
+  @property() public entityFilter?: HaEntityPickerEntityFilterFunc;
+
+  @internalProperty() private _areas?: { [areaId: string]: AreaRegistryEntry };
+
+  @internalProperty() private _devices?: {
+    [deviceId: string]: DeviceRegistryEntry;
+  };
+
+  @internalProperty() private _entities?: EntityRegistryEntry[];
+
+  @internalProperty() private _addMode?: "area_id" | "entity_id" | "device_id";
+
+  @query("#input") private _inputElement?;
+
+  public hassSubscribe(): UnsubscribeFunc[] {
+    return [
+      subscribeAreaRegistry(this.hass.connection!, (areas) => {
+        const areaLookup: { [areaId: string]: AreaRegistryEntry } = {};
+        for (const area of areas) {
+          areaLookup[area.area_id] = area;
+        }
+        this._areas = areaLookup;
+      }),
+      subscribeDeviceRegistry(this.hass.connection!, (devices) => {
+        const deviceLookup: { [deviceId: string]: DeviceRegistryEntry } = {};
+        for (const device of devices) {
+          deviceLookup[device.id] = device;
+        }
+        this._devices = deviceLookup;
+      }),
+      subscribeEntityRegistry(this.hass.connection!, (entities) => {
+        this._entities = entities;
+      }),
+    ];
+  }
+
+  protected render() {
+    if (!this._areas || !this._devices || !this._entities) {
+      return html``;
+    }
+    return html`<div class="mdc-chip-set items">
+        ${ensureArray(this.value?.area_id)?.map((area_id) => {
+          const area = this._areas![area_id];
+          return this._renderChip(
+            "area_id",
+            area_id,
+            area?.name || area_id,
+            undefined,
+            mdiSofa
+          );
+        })}
+        ${ensureArray(this.value?.device_id)?.map((device_id) => {
+          const device = this._devices![device_id];
+          return this._renderChip(
+            "device_id",
+            device_id,
+            device?.name || device_id,
+            undefined,
+            mdiDevices
+          );
+        })}
+        ${ensureArray(this.value?.entity_id)?.map((entity_id) => {
+          const entity = this.hass.states[entity_id];
+          return this._renderChip(
+            "entity_id",
+            entity_id,
+            entity ? computeStateName(entity) : entity_id,
+            entity ? stateIcon(entity) : undefined
+          );
+        })}
+      </div>
+      ${this._renderPicker()}
+      <div class="mdc-chip-set">
```

Comment:
In a future PR, should we make this a component ?

Reply:
I'm waiting for `mwc-chip-set`, but they delayed that again...